edgedb/protocol/codecs/codecs.pyx

```diff
@@ -798,6 +798,61 @@ cdef pgvector_decode(pgproto.CodecContext settings, FRBuffer *buf):
     return val
 
 
+cdef pgsparse_encode(pgproto.CodecContext settings, WriteBuffer buf,
+                     object obj):
+    cdef:
+        int16_t n_elem = 0
+        int16_t dim
+        Py_ssize_t i
+
+    if not _is_array_iterable(obj):
+        raise TypeError(
+            'a sized iterable container expected (got type {!r})'.format(
+                type(obj).__name__))
+
+    # The actual dimentionality of the vector is the size of the raw array
+    dim = len(obj)
+    if dim > PGVECTOR_MAX_DIM:
+        raise ValueError('too many elements in vector value')
+
+    # First pass to count the number of non-zero elements
+    for i in range(dim):
+        if obj[i] != 0:
+            n_elem += 1
+
+    buf.write_int32(6 + n_elem*8)
+    buf.write_int16(n_elem)
+    buf.write_int16(dim)
+    buf.write_int16(0)
+    # Second pass will write the actual non-zero elements
+    for i in range(dim):
+        if obj[i] != 0:
+            buf.write_int32(i)
+            buf.write_float(obj[i])
+
+
+cdef pgsparse_decode(pgproto.CodecContext settings, FRBuffer *buf):
+    cdef:
+        int16_t n_elem
+        int16_t dim
+        Py_ssize_t i
+        int32_t index
```

Comment:
Maybe I don't quite understand some subtle interaction here, but `pg_sparse` code has this when receiving binary data:
```
	int16		n_elem;
	int16		dim;
	int16		unused;

	n_elem = pq_getmsgint(buf, sizeof(int16));
	dim = pq_getmsgint(buf, sizeof(int16));
```

Which, I think, will interpret the value as a signed int16, even though it will eventually assign *that* to an int32 n_elem in the Vector struct. Honestly I find this whole business of jumping between int16 and int32 a lot more odd than whether it's unsigned. 

https://github.com/paradedb/paradedb/blob/585c5dae321fa99576f1399a9589887573e0c700/pg_sparse/src/svector.c#L423-L428